### PR TITLE
Issue/481 crud task templates

### LINKED ISSUE #480: Load personal task templates

### LINKED ISSUE #481: Create, update and delete task templates

### PR DESCRIPTION
**Pull request by florianjs**
## Which issues does this pull request close?
closes #481
closes #480

Hello, in this pull request i added the backend implemenation for Task Template. Task Templates can now be created updated and delated with the backend

Currently there a two functionalities not implentend
* Update Task Template Subtasks
* ward template also shown under personal taks templates

I think we should probably think about including some of that functionality in the backend.

### CODE REVIEW

**Comment by DasProffi** [comment]
added #480 to closes, because this PR already implements it

File: tasks/mutations/task_template_mutations.ts [hunk 1]
```diff
@@ -61,21 +95,22 @@ export const useUpdateMutation = (queryKey: QueryKey, setTemplate: (taskTemplate
...
-    onMutate: async (taskTemplate: TaskTemplateDTO) => {
+    onMutate: async () => {
       await queryClient.cancelQueries({ queryKey: [queryKey] })
       const previousTaskTemplates = queryClient.getQueryData<TaskTemplateDTO[]>([queryKey])
       queryClient.setQueryData<TaskTemplateDTO[]>(
         [queryKey],
-        // TODO do optimistic update here
         (old) => old)
       return { previousTaskTemplates }
     },
```
Comment (MaxSchaefer) [inline tasks/mutations/task_template_mutations.ts:116]: This can be removed

File: tasks/mutations/task_template_mutations.ts [hunk 2]
```diff
@@ -122,21 +165,17 @@ export const useDeleteMutation = (queryKey: QueryKey, setTemplate: (task:TaskTem
...
     onMutate: async (taskTemplate: TaskTemplateDTO) => {
       await queryClient.cancelQueries({ queryKey: [queryKey] })
       const previousTaskTemplate = queryClient.getQueryData<TaskTemplateDTO[]>([queryKey])
       queryClient.setQueryData<TaskTemplateDTO[]>(
         [queryKey],
-        // TODO do optimistic update here
         (old) => old)
       return { previousTaskTemplate }
     },
```
Comment (MaxSchaefer) [inline tasks/mutations/task_template_mutations.ts:181]: This can be removed

File: tasks/mutations/task_template_mutations.ts [hunk 3]
```diff
@@ -91,27 +126,35 @@ export const useUpdateMutation = (queryKey: QueryKey, setTemplate: (taskTemplate
...
+    onError: (_, newTodo, context) => {
+      queryClient.setQueryData([queryKey], context === undefined ? [] : context.previousTaskTemplate)
+    },
+    onMutate: async () => {
       await queryClient.cancelQueries({ queryKey: [queryKey] })
       const previousTaskTemplate = queryClient.getQueryData<TaskTemplateDTO[]>([queryKey])
-      // TODO do optimistic update here
       queryClient.setQueryData<TaskTemplateDTO[]>([queryKey], (old) => old)
       return { previousTaskTemplate }
     },
```
Comment (MaxSchaefer) [inline tasks/mutations/task_template_mutations.ts:157]: This can be removed

**Review by MaxSchaefer — CHANGES_REQUESTED** [review]

File: tasks/mutations/room_mutations.ts [hunk 4]
```diff
@@ -6,7 +6,7 @@ export type TaskStatus = 'unscheduled' | 'inProgress' | 'done'
 
 export type SubTaskDTO = {
   name: string,
-  isDone: boolean
+  isDone?: boolean
```
[the remark this answers is not in the document]
Comment (florianjs) [reply tasks/mutations/room_mutations.ts:9]: I would suggest leaving it at the current optional type.

File: tasks/mutations/task_template_mutations.ts [hunk 5]
```diff
@@ -12,45 +20,71 @@ export type TaskTemplateWardPreviewDTO = {
...
 
 export type TaskTemplateDTO = {
+  wardId? : string,
   id: string,
   name: string,
   notes: string,
-  subtasks: SubTaskDTO[],
+  subtasks: {
+    isDone?: boolean,
+    id?: string,
```
[the remark this answers is not in the document]
Comment (florianjs) [reply tasks/mutations/task_template_mutations.ts:29]: I would suggest leaving it at the current optional type.

File: tasks/mutations/task_template_mutations.ts [hunk 6]
```diff
@@ -12,45 +20,71 @@ export type TaskTemplateWardPreviewDTO = {
...
-  { id: 'id14', subtasks: [], name: 'Ward Template 7', notes: '', isPublicVisible: false }
-]
-
-export const useTaskTemplateQuery = (queryKey: QueryKey, onSuccess: (data: TaskTemplateDTO[]) => void = noop) => {
+type QueryKey = 'personalTaskTemplates' | 'wardTaskTemplates'
+
+export const useWardTaskTemplateQuery = (wardId? : string, onSuccess: (data: TaskTemplateDTO[]) => void = noop) => {
   return useQuery({
-    queryKey: [queryKey],
+    queryKey: ['wardTaskTemplates'],
```
[the remark this answers is not in the document]
Comment (MaxSchaefer) [reply tasks/mutations/task_template_mutations.ts:39]: Sorry. That's a type.

**Pull request closed without merging**